File: desktop/MosDock/QMatrix.cs
using System.Text.RegularExpressions;

namespace MosDock;

readonly record struct QNodeHit(
    string StepId,
    string SkillType,
    string Status,
    string Label,
    string Detail);

static class QMatrix
{
    public const string Nav = "Navigation";
    public const string Tool = "Tool_Usage";
    public const string Cfg = "Configuration";

    static readonly HashSet<string> NavActions = new(StringComparer.OrdinalIgnoreCase)
    {
        "find", "find_navigate", "results_tab", "goto_page", "goto_bookmark",
        "goto_graphic", "goto", "search_query", "search_navigate",
    };

    public static string SkillLabel(string? skill)
    {
        var key = (skill ?? "").Replace(" ", "_");
        return key switch
        {
            Nav or "locate" => "Định vị",
            Tool or "ToolUsage" => "Công cụ",
            Cfg or "Execution" or "configure" => "Tham số",
            _ => string.IsNullOrWhiteSpace(skill) ? "Kỹ năng" : skill,
        };
    }

    public static IReadOnlyList<JsonQNode> NodesFor(JsonCriterion item)
    {
        if (item.QMatrixNodes is { Count: > 0 })
        {
            return item.QMatrixNodes;
        }

        var steps = item.HelpSteps ?? [];
        var pred = item.Predicate?.Type ?? "";
        return
        [
            new JsonQNode
            {
                StepId = "q1",
                SkillType = Nav,
                ValidationRule = "check_cursor_position",
                SuccessMessage = "Đã định vị đúng vị trí.",
                ErrorFeedback = NavFail(steps, item),
            },
            new JsonQNode
            {
                StepId = "q2",
                SkillType = Tool,
                ValidationRule = "check_tool(" + pred + ")",
                SuccessMessage = "Đã gọi đúng công cụ.",
                ErrorFeedback = ToolFail(steps, item),
            },
            new JsonQNode
            {
                StepId = "q3",
                SkillType = Cfg,
                ValidationRule = "check_config(" + pred + ")",
                SuccessMessage = "Đã cấu hình đúng tham số.",
                ErrorFeedback = CfgFail(steps, item),
            },
        ];
    }

    public static LocalCriterion Attach(
        WordFacts facts,
        JsonCriterion item,
        IReadOnlyList<ActionEvent>? evidence,
        LocalCriterion hit)
    {
        var nodes = Evaluate(facts, item, evidence, hit.Status);
        var first = nodes.FirstOrDefault(n => n.Status == "fail");
        if (hit.Status == "fail" && !string.IsNullOrWhiteSpace(first.Detail))
        {
            return hit with
            {
                Message = first.Detail,
                QTrace = nodes,
                BreakSkill = first.SkillType,
            };
        }

        return hit with { QTrace = nodes };
    }

    public static IReadOnlyList<QNodeHit> Evaluate(
        WordFacts facts,
        JsonCriterion item,
        IReadOnlyList<ActionEvent>? evidence,
        string endStatus)
    {
        var events = evidence ?? [];
        var spec = NodesFor(item);
        var outNodes = new List<QNodeHit>();
        var broken = false;
        foreach (var node in spec)
        {
            string status;
            string detail;
            if (endStatus == "error")
            {
                status = "skip";
                detail = First(node.ErrorFeedback, "Không đọc được tệp để chấm bước này.");
            }
            else if (endStatus == "unverified")
            {
                status = "unverified";
                detail = First(node.ErrorFeedback, "Chưa thu được bằng chứng thao tác.");
            }
            else if (endStatus == "pass")
            {
                status = "pass";
                detail = First(node.SuccessMessage, "Đạt.");
            }
            else if (broken)
            {
                status = "skip";
                detail = "Chưa tới bước này.";
            }
            else if (RuleOk(node, item, facts, events))
            {
                status = "pass";
                detail = First(node.SuccessMessage, "Đạt.");
            }
            else
            {
                status = "fail";
                detail = First(node.ErrorFeedback, CfgFail(item.HelpSteps, item));
                broken = true;
            }

            outNodes.Add(new QNodeHit(
                string.IsNullOrWhiteSpace(node.StepId) ? "q" + (outNodes.Count + 1) : node.StepId,
                string.IsNullOrWhiteSpace(node.SkillType) ? Cfg : node.SkillType,
                status,
                SkillLabel(node.SkillType),
                detail));
        }

        return outNodes;
    }

    static bool RuleOk(JsonQNode node, JsonCriterion item, WordFacts facts, IReadOnlyList<ActionEvent> events)
    {
        var rule = (node.ValidationRule ?? "").ToLowerInvariant();
        var skill = node.SkillType ?? "";
        if (rule.Contains("cursor") || rule.Contains("position") || skill == Nav)
        {
            return HasAnchor(facts, item, events);
        }

        if (rule.Contains("count_footnotes"))
        {
            return facts.FootnoteCount >= (item.Predicate.Min ?? 1);
        }

        if (rule.Contains("ribbon") || rule.Contains("tool") || skill == Tool)
        {
            return HasTool(facts, item, events);
        }

        return false;
    }

    static bool HasAnchor(WordFacts facts, JsonCriterion item, IReadOnlyList<ActionEvent> events)
    {
        if (events.Any(ev => NavActions.Contains(ev.Action ?? "")))
        {
            return true;
        }

        var pred = (item.Predicate.Type ?? "").ToLowerInvariant();
        var text = facts.DocumentText ?? "";
        if (pred is "header_contains" or "header_instruction" or "first_page_header" or "watermark_text")
        {
            return facts.HeaderTexts.Count > 0 || facts.HeaderInstructions.Count > 0 || facts.Watermarks.Count > 0 || text.Trim().Length > 0;
        }

        var style = item.Predicate.Style ?? "";
        if (pred == "paragraph_style" && style.StartsWith("TOC", StringComparison.OrdinalIgnoreCase))
        {
            return facts.Fields.Any(f => (f ?? "").Contains("TOC", StringComparison.OrdinalIgnoreCase))
                || facts.StyleCounts.Keys.Any(k => k.StartsWith("TOC", StringComparison.OrdinalIgnoreCase));
        }

        return text.Trim().Length > 0 || facts.Paragraphs.Count > 0 || facts.Ok;
    }

    static bool HasTool(WordFacts facts, JsonCriterion item, IReadOnlyList<ActionEvent> events)
    {
        var pred = (item.Predicate.Type ?? "").ToLowerInvariant();
        if (string.Equals(item.Kind, "action_sequence", StringComparison.OrdinalIgnoreCase))
        {
            return events.Count > 0;
        }

        if (pred == "footnote_min")
        {
            return facts.FootnoteCount >= 1;
        }

        if (pred == "field_contains")
        {
            return facts.Fields.Count > 0;
        }

        if (pred == "paragraph_style")
        {
            var style = item.Predicate.Style ?? "";
            if (style.StartsWith("TOC", StringComparison.OrdinalIgnoreCase))
            {
                return facts.StyleCounts.Keys.Any(k => k.StartsWith("TOC", StringComparison.OrdinalIgnoreCase))
                    || facts.Fields.Any(f => (f ?? "").Contains("TOC", StringComparison.OrdinalIgnoreCase));
            }

            return facts.StyleCounts.Keys.Any(k => k.StartsWith("Heading", StringComparison.OrdinalIgnoreCase) || k == style)
                || facts.Paragraphs.Any(p => (p.Style ?? "").StartsWith("Heading", StringComparison.OrdinalIgnoreCase));
        }

        if (pred == "contains_text")
        {
            var needle = item.Predicate.Text ?? "";
            var blob = facts.DocumentText ?? "";
            var words = Regex.Matches(needle, @"\w+").Select(m => m.Value).Where(w => w.Length > 3).ToArray();
            return words.Length == 0
                ? blob.Trim().Length > 0
                : words.Any(w => blob.Contains(w, StringComparison.OrdinalIgnoreCase));
        }

        if (pred == "not_contains_text")
        {
            return true;
        }

        if (pred.StartsWith("table", StringComparison.Ordinal))
        {
            return facts.Tables.Count > 0;
        }

        return pred switch
        {
            "page_background" => !string.IsNullOrWhiteSpace(facts.PageBackground),
            "watermark_text" => facts.Watermarks.Count > 0,
            "page_border" => facts.PageBorder,
            "header_contains" or "header_instruction" or "first_page_header" =>
                facts.HeaderTexts.Count > 0 || facts.HeaderInstructions.Count > 0,
            "bookmark_range" => facts.Bookmarks.Count > 0,
            "internal_hyperlink" => facts.InternalHyperlinks.Count > 0 || facts.Fields.Count > 0,
            "comment_text" or "comment_author" or "comment_absent_text" or "comment_resolved" or "comment_reply"
                => facts.Comments.Count > 0 || facts.CommentCount > 0,
            "style_used" or "style_size" => facts.StyleCounts.Count > 0,
            "drawing_kind" or "drawing_text" or "artistic_effect" or "picture_effect" or "alt_text" or "wrap_type" or "hdphoto"
                => facts.DrawingKinds.Count > 0 || facts.DrawingTexts.Count > 0 || facts.AltTexts.Count > 0,
            "section_columns" or "page_orientation" or "break_present" or "section_count"
                => facts.Sections.Count > 0 || facts.Breaks.Count > 0,
            "list_format" => facts.NumberingFormats.Count > 0 || facts.Paragraphs.Any(p => !string.IsNullOrWhiteSpace(p.Fmt)),
            _ => false,
        };
    }

    static string NavFail(IReadOnlyList<string> steps, JsonCriterion item) =>
        steps.Count > 0
            ? "Bạn chưa chọn hoặc click đúng vị trí trước khi thao tác. " + Plain(steps[0])
            : "Bạn chưa đặt con trỏ chuột đúng vị trí cho: " + (item.Prompt ?? "nhiệm vụ này") + ".";

    static string ToolFail(IReadOnlyList<string> steps, JsonCriterion item) =>
        steps.Count >= 2
            ? "Bạn đang sử dụng sai công cụ. Thao tác chuẩn: " + Plain(steps[1])
            : "Sai công cụ. Hãy dùng đúng lệnh trên Ribbon cho: " + (item.Prompt ?? "nhiệm vụ này") + ".";

    static string CfgFail(IReadOnlyList<string> steps, JsonCriterion item)
    {
        var extra = steps.Count >= 3 ? Plain(steps[2]) : (item.Feedback?.Fail ?? "");
        return string.IsNullOrWhiteSpace(extra)
            ? "Bạn đã mở đúng công cụ nhưng chọn sai tham số."
            : "Bạn đã mở đúng công cụ nhưng chọn sai tham số hoặc chưa đủ yêu cầu. " + extra;
    }

    static string Plain(string text) => Regex.Replace(text ?? "", @"\*\*", "").Trim();

    static string First(string? a, string b) => string.IsNullOrWhiteSpace(a) ? b : a;
}

File: app/qmatrix.py
"""Q-matrix engine: locate → tool → configure, first break wins."""
from __future__ import annotations

import re

NAV = "Navigation"
TOOL = "Tool_Usage"
CFG = "Configuration"

_NAV_ACTIONS = {
    "find",
    "find_navigate",
    "results_tab",
    "goto_page",
    "goto_bookmark",
    "goto_graphic",
    "goto",
    "search_query",
    "search_navigate",
}


def skill_label(skill: str) -> str:
    key = (skill or "").replace(" ", "_")
    if key in {NAV, "locate"}:
        return "Định vị"
    if key in {TOOL, "ToolUsage"}:
        return "Công cụ"
    if key in {CFG, "Execution", "configure"}:
        return "Tham số"
    return skill or "Kỹ năng"


def nodes_for(criterion: dict) -> list[dict]:
    explicit = [n for n in (criterion.get("q_matrix_nodes") or []) if isinstance(n, dict)]
    if explicit:
        return explicit
    steps = [str(s).strip() for s in (criterion.get("help_steps") or []) if str(s).strip()]
    pred = ((criterion.get("predicate") or {}).get("type") or "").strip()
    return [
        {
            "step_id": "q1",
            "skill_type": NAV,
            "validation_rule": "check_cursor_position",
            "success_message": "Đã định vị đúng vị trí.",
            "error_feedback": _nav_fail(steps, criterion),
        },
        {
            "step_id": "q2",
            "skill_type": TOOL,
            "validation_rule": f"check_tool({pred or 'ribbon'})",
            "success_message": "Đã gọi đúng công cụ.",
            "error_feedback": _tool_fail(steps, criterion),
        },
        {
            "step_id": "q3",
            "skill_type": CFG,
            "validation_rule": f"check_config({pred or 'end_state'})",
            "success_message": "Đã cấu hình đúng tham số.",
            "error_feedback": _cfg_fail(steps, criterion),
        },
    ]


def evaluate(criterion: dict, facts: dict, end_status: str, events: list | None = None) -> list[dict]:
    events = events or []
    nodes = nodes_for(criterion)
    out: list[dict] = []
    broken = False
    for node in nodes:
        if end_status == "error":
            status = "skip"
            detail = node.get("error_feedback") or "Không đọc được tệp để chấm bước này."
        elif end_status == "unverified":
            status = "unverified"
            detail = node.get("error_feedback") or "Chưa thu được bằng chứng thao tác."
        elif end_status == "pass":
            status = "pass"
            detail = node.get("success_message") or "Đạt."
        elif broken:
            status = "skip"
            detail = "Chưa tới bước này."
        else:
            ok = _rule_ok(node, criterion, facts, events)
            if ok:
                status = "pass"
                detail = node.get("success_message") or "Đạt."
            else:
                status = "fail"
                detail = node.get("error_feedback") or _cfg_fail([], criterion)
                broken = True
        out.append(
            {
                "step_id": node.get("step_id") or f"q{len(out) + 1}",
                "skill_type": node.get("skill_type") or CFG,
                "status": status,
                "success_message": node.get("success_message") or "",
                "error_feedback": node.get("error_feedback") or "",
                "detail": detail,
                "label": skill_label(str(node.get("skill_type") or "")),
            }
        )
    return out


def attach(criterion: dict, facts: dict, result: dict, events: list | None = None) -> dict:
    trace = evaluate(criterion, facts, result.get("status") or "", events)
    result = dict(result)
    result["q_matrix"] = trace
    first = next((n for n in trace if n["status"] == "fail"), None)
    if result.get("status") == "fail" and first:
        result["message"] = first.get("detail") or first.get("error_feedback") or result.get("message")
        result["break_skill"] = first.get("skill_type")
        result["break_step"] = first.get("step_id")
    return result


def _nav_fail(steps: list[str], criterion: dict) -> str:
    if steps:
        return "Bạn chưa chọn hoặc click đúng vị trí trước khi thao tác. " + _plain(steps[0])
    return "Bạn chưa đặt con trỏ chuột đúng vị trí cho: " + str(criterion.get("prompt") or "nhiệm vụ này") + "."


def _tool_fail(steps: list[str], criterion: dict) -> str:
    if len(steps) >= 2:
        return "Bạn đang sử dụng sai công cụ. Thao tác chuẩn: " + _plain(steps[1])
    return "Sai công cụ. Hãy dùng đúng lệnh trên Ribbon cho: " + str(criterion.get("prompt") or "nhiệm vụ này") + "."


def _cfg_fail(steps: list[str], criterion: dict) -> str:
    fb = ((criterion.get("feedback") or {}).get("fail") or "").strip()
    extra = _plain(steps[2]) if len(steps) >= 3 else fb
    if extra:
        return "Bạn đã mở đúng công cụ nhưng chọn sai tham số hoặc chưa đủ yêu cầu. " + extra
    return "Bạn đã mở đúng công cụ nhưng chọn sai tham số."


def _plain(text: str) -> str:
    return re.sub(r"\*\*", "", text or "").strip()


def _rule_ok(node: dict, criterion: dict, facts: dict, events: list) -> bool:
    rule = str(node.get("validation_rule") or "").casefold()
    skill = str(node.get("skill_type") or "")
    if "cursor" in rule or "position" in rule or skill == NAV:
        return _has_anchor(facts, criterion, events)
    if "count_footnotes" in rule:
        minimum = int((criterion.get("predicate") or {}).get("min") or 1)
        return int(facts.get("footnote_count") or 0) >= minimum
    if "ribbon" in rule or "tool" in rule or skill == TOOL:
        return _has_tool(facts, criterion, events)
    return False


def _has_anchor(facts: dict, criterion: dict, events: list) -> bool:
    if any(str(ev.get("action") or "").casefold() in _NAV_ACTIONS for ev in events):
        return True
    pred = ((criterion.get("predicate") or {}).get("type") or "").casefold()
    text = facts.get("document_text") or ""
    if pred in {"header_contains", "header_instruction", "first_page_header", "watermark_text"}:
        return bool(
            facts.get("header_texts")
            or facts.get("header_instructions")
            or facts.get("watermarks")
            or text.strip()
        )
    if pred.startswith("toc") or (
        pred == "paragraph_style" and str((criterion.get("predicate") or {}).get("style") or "").upper().startswith("TOC")
    ):
        fields = facts.get("fields") or []
        styles = facts.get("style_counts") or {}
        return any("TOC" in (f or "").upper() for f in fields) or any(
            str(k).upper().startswith("TOC") for k in styles
        )
    return bool(str(text).strip() or facts.get("paragraphs") or facts.get("ok"))


def _has_tool(facts: dict, criterion: dict, events: list) -> bool:
    pred = ((criterion.get("predicate") or {}).get("type") or "").casefold()
    if (criterion.get("kind") or "") == "action_sequence":
        return bool(events)
    if pred == "footnote_min":
        return int(facts.get("footnote_count") or 0) >= 1
    if pred == "field_contains":
        return bool(facts.get("fields"))
    if pred == "paragraph_style":
        style = str((criterion.get("predicate") or {}).get("style") or "")
        counts = facts.get("style_counts") or {}
        paras = facts.get("paragraphs") or []
        if style.upper().startswith("TOC"):
            return any(str(k).upper().startswith("TOC") for k in counts) or any(
                "TOC" in (f or "").upper() for f in facts.get("fields") or []
            )
        return any(str(k).upper().startswith("HEADING") or str(k) == style for k in counts) or any(
            str(p.get("style") or "").upper().startswith("HEADING") for p in paras
        )
    if pred == "contains_text":
        needle = str((criterion.get("predicate") or {}).get("text") or "")
        blob = (facts.get("document_text") or "").casefold()
        words = [w for w in re.findall(r"\w+", needle) if len(w) > 3]
        return any(w.casefold() in blob for w in words) if words else bool(blob.strip())
    if pred == "not_contains_text":
        return True
    if pred.startswith("table"):
        return bool(facts.get("tables"))
    if pred == "page_background":
        return bool(facts.get("page_background"))
    if pred == "watermark_text":
        return bool(facts.get("watermarks"))
    if pred == "page_border":
        return bool(facts.get("page_border"))
    if pred.startswith("header"):
        return bool(facts.get("header_texts") or facts.get("header_instructions"))
    if pred == "bookmark_range":
        return bool(facts.get("bookmarks"))
    if pred == "internal_hyperlink":
        return bool(facts.get("internal_hyperlinks") or facts.get("fields"))
    if pred.startswith("comment"):
        return bool(facts.get("comments") or int(facts.get("comment_count") or 0))
    if pred in {"style_used", "style_size"}:
        return bool(facts.get("style_counts"))
    if pred.startswith("drawing") or pred in {"artistic_effect", "picture_effect", "alt_text", "wrap_type", "hdphoto"}:
        return bool(facts.get("drawing_kinds") or facts.get("drawing_texts") or facts.get("alt_texts"))
    if pred in {"section_columns", "page_orientation", "break_present", "section_count"}:
        return bool(facts.get("sections") or facts.get("breaks"))
    if pred == "list_format":
        return bool(facts.get("numbering_formats")) or any(
            p.get("num_fmt") for p in facts.get("paragraphs") or []
        )
    return False
